juju/juju
# feat(units): wireup facades to get app/unit life from dqlite

Application and unit life is written to dqlite and so can be read from there. The facades which used to get the life value from mongo now use the domain service. There were domain state methods to get life but service methods were needed to expose these to the facades.

As with the EnsureDead and Remove apiserver plugins which relied on a generic mongo FindEntity, the LifeGetter plugin is removed from facades and bespoke code used instead. The code calls the relevant GetUnitLife or GetApplicationLife.

Not all places could be converted. Specifically, there's the life flag worker and similar which will need to be done separately. Also, there's places where more than just unit life is needed and that's not yet in dqlite.

A bit of error cleanup was needed as well, in the area of unit specific errors for already exists, not found.

## QA steps

1. test migration
while units are still coming after after deployment, check the pre checks work, eg
```
$ juju migrate foo c2
ERROR source prechecks failed: unit ubuntu/1 not idle or executing (allocating)
```
after things quiesce, test migration works

2. test k8s scale down
bootstrap to k8s
juju deploy snappass-test -n 3
juju scale-application snappass-test 2
juju destroy-model foo

## Links

**Jira card:** [JUJU-6591](https://warthogs.atlassian.net/browse/JUJU-6591)
